src/components/data-exploration/cell-sets-tool/CellSetsTool.jsx

```diff
@@ -94,45 +98,49 @@ const CellSetsTool = (props) => {
     }
   }, [notifications]);
 
-  const onNodeUpdate = (key, data) => {
+  const [cellSetTreeData, setCellSetTreeData] = useState();
+  const [metadataTreeData, setMetadataTreeData] = useState();
+
+  useEffect(() => {
+    setCellSetTreeData(composeTree(hierarchy, properties, 'cellSets'));
+    setMetadataTreeData(composeTree(hierarchy, properties, 'metadataCategorical'));
+  }, [hierarchy, properties]);
+
+  const [numSelected, setNumSelected] = useState(0);
+
+  useEffect(() => {
+    const selected = allSelected[activeTab];
+    const selectedCells = union(selected, properties);
+
+    const numSelectedUnfiltered = new Set([...selectedCells]
+      .filter((cellIndex) => !filteredCells.current.has(cellIndex)));
+    setNumSelected(numSelectedUnfiltered.size);
+  }, [activeTab, allSelected]);
+
+  const onNodeUpdate = useCallback((key, data) => {
     dispatch(updateCellSetProperty(experimentId, key, data));
-  };
+  }, []);
 
-  const onNodeDelete = (key) => {
+  const onNodeDelete = useCallback((key) => {
```

Comment:
I didn't know about this `useCallback`. Do you know what was the impact on performance when you added it in? We might want to tell others in the team about it

Reply:
We talked with Stefan about it when reviewing their PR.

It is a good idea to tell people about it, although it is something that has to be used with care. In this case, this resulted in some performance improvement as a change to the function called a cascading rerender, including a new cell set tree being generated each time. That took a long time, and hurt performance. However, in most cases, `useCallback` can actually cause a performance penalty.

I can explain more on Slack if you want.